File: apps/claude-plugin/cmd/hooks/validate_session.go
package main

import (
	"encoding/json"
	"io"
	"os"

	validation "github.com/peterkloss/brain/packages/validation"
)

// ValidateSessionInput represents the input for validate-session command
type ValidateSessionInput struct {
	SessionLogPath string `json:"sessionLogPath,omitempty"`
}

// ValidateSessionOutput represents the output for validate-session command
type ValidateSessionOutput struct {
	Valid       bool    `json:"valid"`
	Checks      []Check `json:"checks"`
	Message     string  `json:"message"`
	Remediation string  `json:"remediation,omitempty"`
}

// RunValidateSession handles the validate-session command
func RunValidateSession() error {
	// Read input from stdin (optional)
	input, _ := io.ReadAll(os.Stdin)

	var validateInput ValidateSessionInput
	if len(input) > 0 {
		json.Unmarshal(input, &validateInput)
	}

	// Load workflow state (brain session)
	workflowState, _ := loadWorkflowState()

	// Convert to validation package types
	var state *validation.WorkflowState
	if workflowState != nil {
		state = &validation.WorkflowState{
			Mode:      workflowState.Mode,
			Task:      workflowState.Task,
			SessionID: workflowState.SessionID,
			UpdatedAt: workflowState.UpdatedAt,
		}
	}

	// Run validation using shared package
	result := validation.ValidateSession(state)

	// Convert to output format
	output := ValidateSessionOutput{
		Valid:       result.Valid,
		Message:     result.Message,
		Remediation: result.Remediation,
	}

	for _, c := range result.Checks {
		output.Checks = append(output.Checks, Check{
			Name:    c.Name,
			Passed:  c.Passed,
			Message: c.Message,
		})
	}

	// Output JSON
	if err := outputJSON(output); err != nil {
		return err
	}

	// Exit with appropriate code
	if !result.Valid {
		os.Exit(1)
	}

	return nil
}

File: hooks/scripts/validate-session.ts
/**
 * Validate-session command handler.
 *
 * Ported from apps/claude-plugin/cmd/hooks/validate_session.go.
 * Validates session state for completeness before ending.
 */
import type {
  ValidateSessionInput,
  ValidateSessionOutput,
  WorkflowState,
} from "./types";
import { validateSession } from "./validate";
import { execCommand } from "./exec";

/** Load workflow state from brain session command. */
function loadWorkflowState(): WorkflowState | null {
  try {
    const output = execCommand("brain", ["session", "get-state"]);
    const parsed = JSON.parse(output);
    return {
      mode: parsed.mode ?? parsed.currentMode ?? "",
      task: parsed.task ?? "",
      sessionId: parsed.sessionId ?? "",
      updatedAt: parsed.updatedAt ?? "",
    } satisfies WorkflowState;
  } catch {
    return null;
  }
}

export async function runValidateSession(): Promise<void> {
  const chunks: Buffer[] = [];
  for await (const chunk of process.stdin) {
    chunks.push(chunk as Buffer);
  }
  const raw = Buffer.concat(chunks).toString("utf-8").trim();

  let _validateInput: ValidateSessionInput = {};
  if (raw) {
    try {
      _validateInput = JSON.parse(raw) as ValidateSessionInput;
    } catch {
      // Ignore parse errors
    }
  }

  const workflowState = loadWorkflowState();
  const result = validateSession(workflowState);

  const output: ValidateSessionOutput = {
    valid: result.valid,
    checks: result.checks,
    message: result.message,
    remediation: result.remediation || undefined,
  };

  process.stdout.write(JSON.stringify(output, null, 2) + "\n");

  if (!result.valid) {
    process.exit(1);
  }
}

// Run if executed directly
const isMain = process.argv[1]?.endsWith("validate-session.ts") ||
  process.argv[1]?.endsWith("validate-session.js");
if (isMain) {
  runValidateSession().catch((err) => {
    process.stderr.write(`Error: ${err}\n`);
    process.exit(1);
  });
}
